docs/contracts/v4/guides/07-read-pool-state.mdx

```diff
@@ -2,8 +2,258 @@
+
+```solidity
+contract PoolManager {
+    using Pool for Pool.State;
+    mapping(PoolId => Pool.State) internal pools;
+
+    function swap(PoolId id, ...) external {
+        pools[id].swap(...); // Library call
+    }
+}
+```
+
+This design allows all AMM logic to be encapsulated within the `PoolManager` contract.
+
+## Reading Pool State in v4
+
+In Uniswap v4, reading pool state involves a few key concepts and mechanisms that differ from previous versions. At the core of this new structure is a complex mapping within the PoolManager contract:
+
+```solidity
+mapping(PoolId id => Pool.State) internal _pools;
+```
+
+This mapping represents a fundamental shift in pool data storage:
+
+1. Each pool is identified by a unique `PoolId`.
+2. The `Pool.State` is a struct that contains all the state variables for a single pool.
+3. This struct itself contains several nested mappings and complex data structures.
+
+For example, the `Pool.State` struct might look something like this (simplified for illustration):
+
+```solidity
+struct State {
+    uint160 sqrtPriceX96;
+    int24 tick;
+    uint128 liquidity;
+    uint256 feeGrowthGlobal0X128;
+    uint256 feeGrowthGlobal1X128;
+    mapping(int24 => TickInfo) ticks;
+    mapping(bytes32 => Position.Info) positions;
+    // ... other fields
+}
+```
+
+This complex structure allows for efficient storage of multiple pools and their associated data within a single contract. However, it also means that traditional getter functions would be inefficient or impractical for accessing this data, especially for nested mappings like `ticks` and `positions`.
+
+To address this, Uniswap V4 introduces the StateLibrary and the concept of using `extsload` for reading pool state. These mechanisms provide efficient ways to access the data stored in this complex structure.
+
+### The StateLibrary and `extsload`
+
+```solidity
+abstract contract Extsload is IExtsload {
+    /// @inheritdoc IExtsload
+    function extsload(bytes32 slot) external view returns (bytes32) {
+        assembly ("memory-safe") {
+            mstore(0, sload(slot))
+            return(0, 0x20)
+        }
+    }
+
+    // [...]
+}
+```
+
+The `StateLibrary` is a crucial component in Uniswap v4 for reading pool state. It utilizes the `extsload` function, which is an external wrapper for the `SLOAD` opcode. This allows for efficient reading of arbitrary storage slots.
+
+**How `extsload` works:**
+
+- It takes a storage slot as input.
+- It reads the value stored in that slot directly, using `SLOAD`, from the contract's storage.
+- It returns the value as a `bytes32`.
+
+This method is more gas-efficient than traditional getter functions, especially when reading multiple storage slots.
+
+Moreover, using `extsload` instead of hand-written Solidity view functions helps to keep the contract bytecode size down. This is particularly important for Uniswap v4, as the core contracts are already quite close to Ethereum's contract size limit.
```

Comment:
```suggestion
Moreover, using `extsload` instead of hand-written Solidity view functions lowers the contract bytecode size. This optimization is particularly important for Uniswap v4, as the core contracts are nearly at Ethereum's contract size limit.
```